feat(runtime): implement bidirectional pipe sync and rootless container execution

Alex-Hunterz opened the pull request:
# 🔧 Rootless Container Foundation with Pipe-Based Synchronization

## 📋 Summary

Implements the core foundation for a rootless container runtime in Rust, featuring bidirectional pipe-based IPC to eliminate race conditions during UID/GID mapping and namespace setup.

**Key Result**: Successfully achieves `uid=0(root)` inside container namespace while running unprivileged on host system.

## ✨ What's Implemented

### 🔄 **Bidirectional Pipe IPC**
- Parent↔Child communication channels for atomic coordination
- Eliminates race conditions between namespace creation and UID/GID mapping
- Child signals 'R' (namespace ready), Parent signals 'M' (mapping complete)
- RAII-based file descriptor management with automatic cleanup

### 🔐 **Rootless User Namespace Setup**
- Unprivileged user namespace creation in child process
- Setgroups disabling for secure UID/GID mapping
- Static mapping: host UID 1000 → container root UID 0
- Uses `newuidmap`/`newgidmap` utilities for secure identity mapping

### 🏗️ **Basic Namespace Isolation**
- User namespace (`CLONE_NEWUSER`) - tested and working
- Mount, UTS, PID namespaces (`CLONE_NEWNS | CLONE_NEWUTS | CLONE_NEWPID`) - basic unshare implementation
- **Note**: Full namespace functionality not extensively tested yet

### 🛠️ **Clean Architecture**
- Fork-based process creation with clear parent/child separation
- Consolidated syscalls module for organized low-level operations
- Comprehensive error handling with detailed logging

## 🧪 **Testing Results**

### ✅ **Verified Working**
```bash
$ cargo run -- create --bundle . test
Creating container 'test' with bundle '.'
[Sync] Pipes created:
  Parent->Child: read_fd=3, write_fd=4
  Child->Parent: read_fd=5, write_fd=6
Bento.rs Rootless Container Runtime
Parent PID: 37909 (running as unprivileged user)
[Parent] Child spawned with PID: 37919
[Parent] Waiting for child namespace ready signal...
[Child] Child process started, PID: 37919
[Child] Created user namespace successfully
[Child] Disabled setgroups for safe mapping
[Child] Signaling namespace ready to parent
[Child] Waiting for mapping complete signal...
[Parent] Received namespace ready signal from child
[Parent] Starting rootless mapping for child 37919
[Parent] Host UID: 1000, Host GID: 1000
[Parent] Executing: newgidmap 37919 0 1000 1
[Parent] newgidmap succeeded
[Parent] Executing: newuidmap 37919 0 1000 1
[Parent] newuidmap succeeded
[Parent] Rootless mapping complete: host user -> container root
[Parent] Mapping complete
[Parent] Signaled mapping complete to child
[Child] Received mapping complete signal
[Child] Created remaining namespaces: CloneFlags(CLONE_NEWNS | CLONE_NEWUTS | CLONE_NEWPID)
[Child] Executing command: ["/usr/bin/id"]
uid=0(root) gid=0(root) groups=0(root),65534(nogroup)
```

🎯 Technical Implementation
Synchronization Flow

```text
1. Parent creates bidirectional pipes
2. Fork creates parent and child processes  
3. Child creates user namespace, disables setgroups
4. Child signals 'R' → Parent (namespace ready)
5. Parent performs UID/GID mapping with newuidmap/newgidmap
6. Parent signals 'M' → Child (mapping complete)
7. Child creates remaining namespaces and executes command
```
Architecture

```text
crates/libbento/src/
├── process.rs    # Container creation with pipe synchronization
└── syscalls.rs   # Consolidated syscall wrappers
    ├── Process creation (fork_intermediate)
    ├── Namespace ops (unshare_user_namespace, unshare_remaining_namespaces)  
    └── UID/GID mapping (map_user_namespace_rootless, disable_setgroups_for_child)
```

Alex-Hunterz committed:
feat(runtime): implement bidirectional pipe sync and rootless container execution

- Add bidirectional pipe-based IPC for race-free parent-child synchronization
- Implement fork-based process creation with clean role separation
- Create user namespaces in child and disable setgroups for secure mapping
- Perform UID/GID mapping using newuidmap/newgidmap utilities from parent
- Setup remaining namespaces (mount, UTS, PID) after successful identity mapping
- Achieve container root privileges without requiring host root access
- Eliminate race conditions through explicit handshake protocol ('R'/'M' signals)
- Progress from initial prototype through unidirectional to bidirectional sync

This establishes the core foundation for Bento.rs as a production-ready rootless,
daemonless, OCI-compliant container runtime with robust synchronization guarantees.

Tested with successful container creation showing uid=0(root) inside namespace
while maintaining unprivileged operation on host system.

Alex-Hunterz commented:
TLDR
currently in the feat/fork-clone branch we have
1. Set up pipes for IPC
2. Forking to create child process
3. setting up user namespaces in child process
4. sync using pipes
5. UID/GID mappping for single user within container. thus providing it root privilidges within

NavneetNayak commented on crates/libbento/src/syscalls.rs:
```diff
@@ -0,0 +1,167 @@
+        match execvp(cmd, &ref_args) {
+            Ok(_) => 0, // Success (execvp replaces process, so this won't run)
+            Err(e) => {
+                eprintln!("execvp failed: {}", e);
```
this is not used?

NavneetNayak commented on crates/libbento/src/process.rs:
```diff
@@ -0,0 +1,168 @@
+            return 1;
+        }
+        
+        // Phase 3: Execute target command
```
move to to a function which the closure calls its cleaner, for both parent and child code, also call them something more representative of their purpose than 'parent' and 'child'

NavneetNayak commented on crates/libbento/src/process.rs:
Also when you do config parsing the parsed types are probably not going to be like the config you've defined here, so keep in mind you'll have to refactor this later

Alex-Hunterz replied:
Yup, will remember this.

NavneetNayak commented on crates/bento-cli/src/main.rs:
You can add the todo()! macro wherever stuff is not yet implemented

Alex-Hunterz replied:
Yes. will keep that in mind.

Alex-Hunterz committed:
refactor: Made the closures cleaner and extracted them to named functions

- Added todo!() macros
- Removed the unused execvp code
- Defined functions to extract the parent_logic and child_logic closures
